### ISSUE TITLE
New function to set a define's default if not already defined

### ISSUE THREAD

**Issue by proseLA** [zencart/zencart#4825]
this section of code is something new in v158 in that we are checking if the config key is already defined:

https://github.com/zencart/zencart/blob/5ab07980bdd846ce174ad976456686a339ecf1f5/laravel/app/Models/Configuration.php#L29-L31

looking in v157, we see we do not have that check:

https://github.com/zencart/zencart/blob/742b837868a2b128df96b6635c413c8187972679/includes/init_includes/init_db_config_read.php#L23-L27

in seeing all of the php 8.0 notices for already defined constants, i think it would behoove us to provide an early loading function to do this.  ie, something like:

```
        function defineIfNot(string $name, string $definition)
        {
            if (!defined($name)) {
                define($name, $definition);
            }
        }
```
i see potential for plenty of uses as store owners move to php 8.0 and are inundated with log files.

thoughts?

noted, that i have the declared variable arguments as both strings and in some instances integer is required.  also noting that i WOULD love to refactor all of those boolean strings into actual booleans... and noting that will never happen.  a coder can dream...

**Comment by scottcwilson** [comment]
If you are using Zen Cart 1.5.7 and PHP 8, you really want to set logging (Report All Errors)  to use IgnoreDups (or No).  Then you won't inundated with log files.

**Comment by lat9** [comment]
> this section of code is something new in v158 in that we are checking if the config key is already defined:
> 
> https://github.com/zencart/zencart/blob/5ab07980bdd846ce174ad976456686a339ecf1f5/laravel/app/Models/Configuration.php#L29-L31
> 
> looking in v157, we see we do not have that check:
> 
> https://github.com/zencart/zencart/blob/742b837868a2b128df96b6635c413c8187972679/includes/init_includes/init_db_config_read.php#L23-L27
> 
> in seeing all of the php 8.0 notices for already defined constants, i think it would behoove us to provide an early loading function to do this. ie, something like:
> 
> ```
>         function defineIfNot(string $name, string $definition)
>         {
>             if (!defined($name)) {
>                 define($name, $definition);
>             }
>         }
> ```
> 
> i see potential for plenty of uses as store owners move to php 8.0 and are inundated with log files.
> 
> thoughts?
> 
> noted, that i have the declared variable arguments as both strings and in some instances integer is required. also noting that i WOULD love to refactor all of those boolean strings into actual booleans... and noting that will never happen. a coder can dream...

PHP 8.0+ [introduced](https://php.watch/versions/8.0/mixed-type) the `mixed` argument type, but that can't be used since zc158 will support PHP versions back to 7.3.

I'm liking the suggestion, but would suggest a renaming so that it's more obvious that the function is one in the Zen Cart base, e.g.
```
function zen_define_default(string $name, $value)
{
    if (!defined($name)) {
        define($name, $value);
    }
}
```

**Comment by proseLA** [comment]
i have zero problem with the name change.  makes sense to me.

tangentially related, i'm liking that addition in the docs.

**Comment by proseLA** [comment]
> If you are using Zen Cart 1.5.7 and PHP 8, you really want to set logging (Report All Errors) to use IgnoreDups (or No). Then you won't inundated with log files.

i am glad that @lat9 saw the value here.  the problem that this addresses is now in the documentation:

In Zen Cart 1.5.8, the language file structure was modified so that the PHP `define` statement is no longer used directly to set a language constant. The reason for this is that newer versions of PHP will emit warning messages when a constant is redefined (in an override vs core file), so the old approach would create debug logs if something wasn't done.

in my experience i was seeing these errors even with ignoreDupes set as logging.

**Comment by scottcwilson** [comment]
Others have reported this too - not sure why it happens on some installations and not others. 
Well, the good news is that it's fixed in 1.5.8.
